src/ImageSharp/Memory/Allocators/Internals/UnmanagedBuffer{T}.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Buffers;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace SixLabors.ImageSharp.Memory.Internals
+{
+    /// <summary>
+    /// Allocates and provides an <see cref="IMemoryOwner{T}"/> implementation giving
+    /// access to unmanaged buffers.
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    internal sealed class UnmanagedBuffer<T> : MemoryManager<T>
+        where T : struct
+    {
+        private bool isDisposed;
+        private readonly SafeHandle safeHandle;
+        private readonly int length;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UnmanagedBuffer{T}"/> class.
+        /// </summary>
+        /// <param name="byteCount">The number of bytes to allocate.</param>
+        public UnmanagedBuffer(int byteCount)
+        {
+            this.length = byteCount / Unsafe.SizeOf<T>();
+            this.safeHandle = new SafeHGlobalHandle(byteCount);
+        }
+
+        public override unsafe Span<T> GetSpan()
+            => new Span<T>((void*)this.safeHandle.DangerousGetHandle(), this.length);
+
+        /// <inheritdoc />
+        public override unsafe MemoryHandle Pin(int elementIndex = 0)
+        {
+            // Will be released in Unpin
+            bool unused = false;
+            this.safeHandle.DangerousAddRef(ref unused);
+
+            void* pbData = Unsafe.Add<T>((void*)this.safeHandle.DangerousGetHandle(), elementIndex);
+            return new MemoryHandle(pbData);
+        }
+
+        /// <inheritdoc />
+        public override void Unpin() => this.safeHandle.DangerousRelease();
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (this.isDisposed || this.safeHandle.IsInvalid)
+            {
+                return;
+            }
+
+            if (disposing)
+            {
+                this.safeHandle.Dispose();
+            }
+
+            this.isDisposed = true;
+        }
+
+        private sealed class SafeHGlobalHandle : SafeHandle
+        {
+            private readonly int byteCount;
+
+            public SafeHGlobalHandle(int size)
+                : base(IntPtr.Zero, true)
+            {
+                this.SetHandle(Marshal.AllocHGlobal(size));
+                GC.AddMemoryPressure(this.byteCount = size);
```

Comment:
Should this be done? This would lead to a faster gen2 collection and object itself would behave like it's a managed allocation which I assume is not what we want to be happening. 

Microsoft docs states that ["The AddMemoryPressure and RemoveMemoryPressure methods improve performance only for types that exclusively depend on finalizers to release the unmanaged resources"](https://docs.microsoft.com/en-us/dotnet/api/system.gc.addmemorypressure?view=net-5.0). Which is not the case here as safe handle is used to prevent memory leaking only when user forgot to call dispose which is very unlikely to be caused by library code (this should be tested somehow to be honest).

Otherwise this would lead to a memory throttle as it is with current implementation.

Reply:
That advice is awkwardly worded.  In the case where a user fails to Dispose an object backed by an unmanaged buffer, it does rely on the SafeHandle's finalizer to release the memory, so it is advantageous to advise the GC that it might be able to reclaim the memory by doing its GC thing.  This should only result in more frequent gen2 GCs if the memory limits are being reached, which should only happen if the unmanaged buffer is long-lived or if the system is actually low on memory.  For properly Disposed ephemeral buffers, the GC will be aware of the added memory pressure but will also see it freed quickly, so no harm.

Reply:
Not sure that these buffers can be named 'ephemeral', 2mb is a lot (at least it looks like a lot for something temporal) which *should* lead to a lot of execution time spent on this memory. Primary idea I had while writing this is that even if user forgets to dispose memory backed by a handle it'll be freed by either gen0/gen1 collection or full gen2 collection when `OutOfMemoryException` kicks in.

While I'm not so sure about this now, I'm still concerned that this might be an overkill due to freachable queue. At least I'd want to benchmark this on something big like your parallel beeheads demo.

Reply:
unmanaged buffers will only be used for allocations that are over the pooled size limit.  They should be used once and then released, hence 'ephemeral'.  The question of whether `GC.AddMemoryPressure` is appropriate ultimately comes down to "is there any way a GC could free this memory?".  The answer here is yes, but only if the memory has been leaked because someone didn't dispose.  Unfortunately there's no way to know whether someone is *going to* leak; you can only tell after they've done it, when your finalizer runs.

What the MS docs should have said is something along the lines of "don't tell the GC about memory it couldn't possibly reclaim".  If the allocation and free were always 100% deterministic, there would be no point in telling the GC about them.  But since we can't know whether the GC *could* reclaim the memory, it's better to tell it about the allocation than to not.

Reply:
Once again, thanks for a deep dive! I've never looked at GC from "memory it can or cannot not claim" point of view.